core/src/main/scala/org/apache/spark/BarrierTaskContext.scala

```diff
@@ -117,12 +118,30 @@ class BarrierTaskContext private[spark] (
     timer.schedule(timerTask, 60000, 60000)
 
     try {
-      barrierCoordinator.askSync[Unit](
+      val cancelableRpcFuture = barrierCoordinator.askCancelable[Unit](
         message = RequestToSync(numTasks, stageId, stageAttemptNumber, taskAttemptId,
           barrierEpoch),
         // Set a fixed timeout for RPC here, so users shall get a SparkException thrown by
         // BarrierCoordinator on timeout, instead of RPCTimeoutException from the RPC framework.
         timeout = new RpcTimeout(365.days, "barrierTimeout"))
+
+      // Wait the RPC future to be completed, but every 1 second it will jump out waiting
+      // and check whether current spark task is killed. If killed, then throw
+      // a `TaskKilledException`, otherwise continue wait RPC until it completes.
+      while(!taskContext.isCompleted()) {
+        if (taskContext.isInterrupted()) {
+          val reason = taskContext.getKillReason().get
```

Comment:
is it possible that `getKillReason` is none while `taskContext.isInterrupted` is true?

Reply:
No. See `isInterrupted` implementation:
```
def isInterrupted(): Boolean = reasonIfKilled.isDefined
```